sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py

```diff
@@ -417,11 +427,16 @@ def _check_categories(self, X):
                 if use_pandas_categorical:
                     # pandas categories do not include missing values so there is
                     # no need to filter them out.
-                    categories = X.iloc[:, f_idx].unique().dropna().to_numpy()
-                    # OrdinalEncoder requires categories backed by numerical values
-                    # to be sorted
-                    if categories.dtype.kind not in "OUS":
-                        categories = np.sort(categories)
+
+                    # relying on the _col attribute is how pandas converts
+                    # interchange Columns up to 2.1.3:
+                    # https://github.com/pandas-dev/pandas/blob/1c606d5f014c5296d6028af28001311b67ee3721/pandas/core/interchange/from_dataframe.py#L211 # noqa
+                    categories = np.sort(
+                        X.__dataframe__()
+                        .get_column(f_idx)
+                        .describe_categorical["categories"]
+                        ._col
```

Comment:
relying on the fact that both pandas and polars happen to store the original Series in `._col` is definitely not great. It is the strategy used in `pandas.from_dataframe` for categorical columns. pyarrow has a [buffers_to_array function](https://github.com/apache/arrow/blob/76f5976147883fa42dcfb6c775073cce3eaf3606/python/pyarrow/interchange/from_dataframe.py#L345) that can actually load the buffers from a dataframe api interchange Column

another option would be to not rely on the dataframe interchange protocol to find the categories, but only to discover the categorical columns. Those can then be loaded from `X` with `np.asarray` and then we can remove missing values and duplicates in a way similar to what is done below when `X` is not a dataframe.
the code would be different than what is done for non-dataframe X because that only works for numbers (eg it relies on `np.isnan`) but polars categories are always strings

Reply:
Indeed, this is quite suboptimal. Decoding the categories array from the raw buffers seems non-trivial with numpy.

Maybe @jorisvandenbossche or @MarcoGorelli have a better suggestion?

Reply:
or maybe we could use the OrdinalEncoder with categories="auto", and check that the number of categories is smaller than `max_bins` after doing the preprocessing, by checking its `categories_` fitted attribute?

otherwise I have the impression we implement the logic for finding categories and handling missing values twice, once here and once in the OrdinalEncoder.

Reply:
in this case here we would just need the interchange protocol to discover which columns are categorical. and rather than implementing it in the gradient boosting maybe the `compose.make_column_selector` could be generalized a bit to allow selecting by interchange `DTypeKind`, so that the preprocessing done in the gradient boosting would rely more on ColumnTransformer, OrdinalEncoder and make_column_selector

Reply:
That sounds like a good plan to me. @thomasjpfan any opinion?

Reply:
Improving `make_column_selector` to select the categories makes sense. I would open another PR to see what it looks like in code.

> maybe we could use the OrdinalEncoder with categories="auto",

This would most likely work. Originally, I used `pandas`'s `unique` function to find the unique categories, because it was faster than having `OrdinalEncoder` automatically do it. In the end, I wanted to updated `OridinalEncoder` to have a fast path for pandas series.

Note, the issue with grabbing the categories from the dtype, is that it is a superset of the actual categories. For example, the input data could consistent of 2 unique categories, but the categorical dtype can consistent of 300 categories. `OridinalEncoder`'s current behavior is only to discover the categories that are actually in the data.

```python
from pandas.api.types import CategoricalDtype
import pandas as pd
from sklearn.preprocessing import OrdinalEncoder

X = pd.DataFrame(
    {
        "f1": pd.Series(["c1", "c2"], dtype=CategoricalDtype(categories=[f"c{i}" for i in range(300)]))
    }
)

enc = OrdinalEncoder().fit(X)
print(enc.categories_)
# [array(['c1', 'c2'], dtype=object)]
```

Also, if there are only 2 actual categories but the dtype has 300 categories, then I expect histogram gradient boosting should still work because 2 actual categories < `max_bins`.

Reply:
that's a good point and another reason for not relying on the dtype but on the actual column values.


> Also, if there are only 2 actual categories but the dtype has 300 categories, then I expect histogram gradient boosting should still work because 2 actual categories < max_bins.

Indeed that makes more sense. A similar issue I noticed is that in the main branch if a categorical column contains values greater than `max_bins` (_before_ re-encoding them with the `OrdinalEncoder`), [this check](https://github.com/scikit-learn/scikit-learn/blob/b8229daafee0e50690d4b8447f93cf1069ba6880/sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py#L484) raises a `ValueError`. However as they get re-encoded later, I think what matters is the number of actual categories (the number of unique values) rather than the max of the column.

```python
import numpy as np
from sklearn.ensemble import HistGradientBoostingRegressor

X = np.asarray([0, 1, 1000])[:, None]
y = [0., 0., 0.]
gb = HistGradientBoostingRegressor(categorical_features=[0])
gb.fit(X, y)
# ValueError: Categorical feature at index 0 is expected to be encoded with values < 255 but the largest value for the encoded categories is 1000.
```

If I remove the check fit and predict work as expected.

Reply:
> that's a good point and another reason for not relying on the dtype but on the actual column values.

At the same time, I find that having dimensions of model parameters, shape of the output of `transform` or `predict_proba` depend on the observed values in `X` or `y` is a bit annoying, especially when trying to consolidate models fit on resampled datasets with rare categories in features or classes in the target. I recently faced the y-variant of this issue when experimenting with a new kind of meta estimator for quantile regression:

https://nbviewer.org/github/ogrisel/notebooks/blob/master/quantile_regression_as_classification.ipynb

In the `predict_quantiles` method, I have to post-process the output of `predict_proba` to deal with unobserved classes.

When the scikit-learn API was designed, we only had numpy arrays in mind. The expected number of classes (and later categories in features) was never present in numpy dtypes, hence we naturally had to infer it implicitly from the training set values.

But now that data science is moving more towards dataframe with richer dtype metadata, maybe this design decision is counter productive.

For classes / `predict_proba`, I think we don't want to change the current behavior, at least not in the short term.

But when adding dataframe support for categorical feature aware transformers / estimators, maybe we should have an option to collapse categories to the observed ones or preserve all categories specified in the dtype, even when unused.

Reply:
Indeed it can be annoying to need to perform this kind of post-processing. Although for the specific case of categorical features (not outputs) for the gradient boosting, I don't see which public attributes or method outputs would be changed by taking into account categories that do not show up in the training data?